oviewer/convert_es.go

```diff
@@ -169,193 +181,292 @@ func (es *escapeSequence) convert(st *parseState) bool {
+// 38;5;n or 38;2;n;n;n semi-colon separated.
+// 38:5:n or 38:2:n:n:n colon separated.
+// 38:2::n:n:n colon separated with double colon.
+//
+// The first return value is the color name.
+// The second return value is the number of parameter index increments.
+// The last return value is an error.
+// Illegal characters will result in an error, while others will be ignored without error.
+func convertSGRColor(sgr sgrParams) (string, int, error) {
+	if len(sgr.params) == 0 {
+		return "", 0, nil
+	}
+	inc := 1
+	ex, err := sgrNumber(sgr.params[0])
 	if err != nil {
-		return 1, ""
+		return "", inc, err
 	}
 	switch ex {
-	case 5: // 8-bit colors.
-		if len(fields) < 3 {
-			return 1, ""
-		}
-		if fields[2] == "" {
-			return len(fields), ""
+	case Colors256:
+		if len(sgr.params) < 2 {
+			return "", inc, nil
 		}
-
-		color, err := parse8BitColor(fields[2])
+		color, err := parse256Color(sgr.params[1])
 		if err != nil {
-			return 0, color
+			return color, inc, err
 		}
-		return 2, color
-	case 2: // 24-bit colors.
-		if len(fields) < 5 {
-			return len(fields), ""
+		inc++
+		return color, inc, nil
+	case ColorsRGB:
+		if len(sgr.params) < 4 {
+			return "", len(sgr.params), nil
 		}
-		for i := 2; i < 5; i++ {
-			if fields[i] == "" {
-				return i, ""
-			}
+		rgb := sgr.params[1:4] // 38:2:r:g:b
+		// The colon(colonF) parameter allows two colons(::) to be done before the RGB is specified.
+		if sgr.colonF && sgr.params[1] == "" && len(sgr.params) > 4 {
+			rgb = sgr.params[2:5] // 38:2::r:g:b
 		}
-
-		color, err := parseRGBColor(fields[2:5])
+		color, err := parseRGBColor(rgb)
 		if err != nil {
-			return 0, color
+			return color, inc, err
 		}
-		return 4, color
+		inc += 3
+		return color, inc, nil
 	}
-	return 1, ""
+	return "", inc, nil
 }
 
-func parse8BitColor(field string) (string, error) {
-	c, err := strconv.Atoi(field)
+// parse256Color parses the 8-bit color.
+func parse256Color(param string) (string, error) {
+	if param == "" {
+		return "", nil
+	}
+	c, err := sgrNumber(param)
 	if err != nil {
-		return "", fmt.Errorf("invalid 8-bit color value: %v", field)
+		return "", err
 	}
 	color := colorName(c)
 	return color, nil
 }
 
-func parseRGBColor(fields []string) (string, error) {
-	red, err1 := strconv.Atoi(fields[0])
-	green, err2 := strconv.Atoi(fields[1])
-	blue, err3 := strconv.Atoi(fields[2])
-	if err1 != nil || err2 != nil || err3 != nil {
-		return "", fmt.Errorf("invalid RGB color values: %v, %v, %v", fields[0], fields[1], fields[2])
-	}
-	color := fmt.Sprintf("#%02x%02x%02x", red, green, blue)
-	return color, nil
-}
-
 // colorName returns a string that can be used to specify the color of tcell.
 func colorName(colorNumber int) string {
 	if colorNumber < 0 || colorNumber > 255 {
 		return ""
 	}
 	return tcell.PaletteColor(colorNumber).String()
 }
+
+// parseRGBColor parses the RGB color.
+func parseRGBColor(params []string) (string, error) {
+	if params[0] == "" || params[1] == "" || params[2] == "" {
```

Comment:
Either you use


```suggestion
func parseRGBColor(params [3]string) (string, error) {
	if params[0] == "" || params[1] == "" || params[2] == "" {
```

Either you return an error `if len(params) != 3)`

Otherwise, panic

Reply:
Thank you.